File: 2019_kakao_blind_recruitment_open_chat_room.py
def solution(record):
    answer = []
    userInfo = {}
    entered = []
    for r in record:
        action, *other = r.split(' ')
        if action == 'Enter':
            user_id, nick = other
            userInfo[user_id] = nick
            answer.append("님이 들어왔습니다.")
            entered.append(user_id)
        elif action == 'Change':
            user_id, nick = other
            userInfo[user_id] = nick
        else:
            user_id = other[0]
            answer.append("님이 나갔습니다.")
            entered.append(user_id)
    
    for i in range(len(answer)):
        answer[i] = userInfo[entered[i]] + answer[i]
    
    return answer

File: 2019_kakao_blind_recruitment_open_chat_room.js
function solution(record) {
  const answer = [];
  const actions = [];
  const uids = [];
  const userInfo = {};

  record.forEach((r) => {
    const [action, uid, nick] = r.split(" ");
    switch (action) {
      case "Enter":
        actions.push("님이 들어왔습니다.");
        userInfo[uid] = nick;
        uids.push(uid);
        break;
      case "Change":
        userInfo[uid] = nick;
        break;
      case "Leave":
        actions.push("님이 나갔습니다.");
        uids.push(uid);
        break;
    }
  });

  actions.forEach((action, i) => {
    answer.push(userInfo[uids[i]] + action);
  });

  return answer;
}
